fw/thread.go

```diff
@@ -221,39 +221,44 @@ func (t *Thread) processIncomingInterest(pendingPacket *ndn.PendingPacket) {
 	}
 
 	// Check if any matching PIT entries (and if duplicate)
-	pitEntry, isDuplicate := t.pitCS.FindOrInsertPIT(interest, fhName, incomingFace.FaceID())
+	pitEntry, isDuplicate := t.pitCS.InsertInterest(interest, fhName, incomingFace.FaceID())
 	if isDuplicate {
 		// Interest loop - since we don't use Nacks, just drop
 		core.LogInfo(t, "Interest ", interest.Name(), " is looping - DROP")
 		return
 	}
-	core.LogDebug(t, "Found or updated PIT entry for Interest=", interest.Name(), ", PitToken=", uint64(pitEntry.Token))
+	core.LogDebug(t, "Found or updated PIT entry for Interest=", interest.Name(), ", PitToken=", uint64(pitEntry.Token()))
 
 	// Get strategy for name
 	strategyName := table.FibStrategyTable.LongestPrefixStrategy(interest.Name())
 	strategy := t.strategies[strategyName.String()]
 	core.LogDebug(t, "Using Strategy=", strategyName, " for Interest=", interest.Name())
 
 	// Add in-record and determine if already pending
-	_, isAlreadyPending := pitEntry.FindOrInsertInRecord(interest, incomingFace.FaceID(), incomingPitToken)
+	_, isAlreadyPending := pitEntry.InsertInRecord(interest, incomingFace.FaceID(), incomingPitToken)
 	if !isAlreadyPending {
 		core.LogTrace(t, "Interest ", interest.Name(), " is not pending")
 
 		// Check CS for matching entry
 		if t.pitCS.IsCsServing() {
-			csEntry := t.pitCS.FindMatchingDataCS(interest)
+			csEntry := t.pitCS.FindDataExactMatch(interest)
 			if csEntry != nil {
-				// Pass to strategy AfterContentStoreHit pipeline
-				strategy.AfterContentStoreHit(pitEntry, incomingFace.FaceID(), csEntry.Data)
-				return
+				csEntry = t.pitCS.FindDataPrefixMatch(interest)
+				if csEntry != nil {
+					// Pass to strategy AfterContentStoreHit pipeline
+					strategy.AfterContentStoreHit(pitEntry, incomingFace.FaceID(), csEntry.Data())
+					return
+				}
+
 			}
 		}
 	} else {
 		core.LogTrace(t, "Interest ", interest.Name(), " is already pending")
 	}
 
 	// Update PIT entry expiration timer
-	pitEntry.UpdateExpirationTimer()
+	table.UpdateExpirationTimer(pitEntry)
```

Comment:
`pitEntry` used to be a struct, but since it's been changed to an interface now, we can no longer call member functions associated with it. Thus, I changed `UpdateExpirationTimer` to take an argument of type `PitEntry` instead. This was the only way I could think of doing this.

Reply:
You can expose this function if necessary. I.e. add to interface.

Reply:
Add `UpdateExpirationTimer` to `PitEntry` interface. I don't know if it will break anything but looks good to me.

Reply:
If we add `UpdateExpirationTimer()` to the `PitEntry` interface, then each individual implementation of the PitCS will need to implement this method separately. This is because `UpdateExpirationTimer()` requires accessing the underlying PitCS table. Specifically, this line: https://github.com/named-data/YaNFD/blob/master/table/pit-cs.go#L355

However, this seems like code that should be kept common to all the PitCS table implementations. However, because of the need to access the underlying concrete PitCS table, it can't be implemented by `basePitEntry`. Thus, that's why I decided to change `UpdateExpirationTimer()` to be a function not associated with any struct.

The following code does not compile:
```

// PitEntry dictates what entries in a PIT-CS table should implement
type PitEntry interface {
	PitCs() PitCsTable
<snipped>
	UpdateExpirationTimer()
	waitForPitExpiry()
}
// UpdateExpirationTimer updates the expiration timer to the latest expiration time of any in or out record in the entry.
func (e *basePitEntry) UpdateExpirationTimer() {
	e.SetExpirationTime((time.Unix(0, 0)))

	for _, record := range e.InRecords() {
		if record.ExpirationTime.After(e.ExpirationTime()) {
			e.SetExpirationTime(record.ExpirationTime)
		}
	}

	for _, record := range e.OutRecords() {
		if record.ExpirationTime.After(e.ExpirationTime()) {
			e.SetExpirationTime(record.ExpirationTime)
		}
	}

	go e.waitForPitExpiry()
}

func (e *basePitEntry) waitForPitExpiry() {
	if !e.ExpirationTime().IsZero() {
		time.Sleep(e.ExpirationTime().Sub(time.Now().Add(time.Millisecond * 1))) // Add 1 millisecond to ensure *after* expiration
		if e.ExpirationTime().Before(time.Now()) {
			// Otherwise, has been updated by another PIT entry
			e.PitCs().ExpiringPitEntries() <- e     // HERE: e.PitCs() fails to compile because basePitEntry cannot implement PitCs()
		}
	}
}
```

Reply:
OK. Then let's keep what you have.